src/boot/efi/boot.c

```diff
@@ -2311,6 +2353,111 @@ static VOID config_write_entries_to_variable(Config *config) {
 }
 
+static EFI_STATUS secure_boot_enroll_from(Config *config, EFI_FILE_HANDLE root_dir, CHAR16 *keys_dir) {
+        static const UINT32 sb_vars_opts =
+                EFI_VARIABLE_NON_VOLATILE |
+                EFI_VARIABLE_BOOTSERVICE_ACCESS |
+                EFI_VARIABLE_RUNTIME_ACCESS |
+                EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS |
+                0;
+
+        static const struct {
+                const CHAR16 *name;
+                const CHAR16 *filename;
+                const EFI_GUID vendor;
+        } sb_vars[] = {
+                { L"db",  L"db.auth",  EFI_IMAGE_SECURITY_DATABASE_VARIABLE },
+                { L"KEK", L"KEK.auth", EFI_GLOBAL_VARIABLE },
+                { L"PK",  L"PK.auth",  EFI_GLOBAL_VARIABLE },
+        };
+
+        EFI_STATUS err = EFI_SUCCESS;
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE dir = NULL;
+
+        err = uefi_call_wrapper(root_dir->Open, 5, root_dir, &dir, keys_dir, EFI_FILE_MODE_READ, 0ULL);
+        if (EFI_ERROR(err)) {
+                Print(L"Failed getting a handle on keys directory %s.\n", keys_dir);
+                return err;
+        }
+
+        for (UINTN i = 0; i < ELEMENTSOF(sb_vars); i++) {
+                _cleanup_freepool_ CHAR8 *buffer = NULL;
+                UINTN size;
+
+                err = file_read(dir, sb_vars[i].filename, 0, 0, &buffer, &size);
+                if (EFI_ERROR(err)) {
+                        Print(L"Failed reading %s file.\n", sb_vars[i].filename);
+                        return err;
+                }
+
+                err = efivar_set_raw(&sb_vars[i].vendor, sb_vars[i].name, buffer, size, sb_vars_opts);
+                if (EFI_ERROR(err)) {
+                        Print(L"Failed to write %s secure boot variable.\n", sb_vars[i].name);
+                        return err;
+                }
+        }
+
+        /* if we loaded the signing keys successfully then we remove all the
+         * signing keys entries from  the menu as the system is now locked down. */
+        config_remove_by_type(config, LOADER_SIGNATURE_KEYS);
+
+        return err;
+}
+
+static EFI_STATUS secure_boot_discover_keys(Config *config, EFI_FILE *root_dir) {
+        EFI_STATUS err = EFI_SUCCESS;
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE keys_basedir = NULL;
+
+        if (!setup_mode_enabled())
+                return EFI_SUCCESS;
+
+        /* if auto enrollement is activated (this is the default), we try to
+         * load keys straight from the /loader/keys/auto location. If it fails 
+         * (e.g. there is no such directory), we still proceed to add the
+         * entries to the menu (including the auto one)
+         */
+        if (config->auto_enroll == true) {
+                err = secure_boot_enroll_from(config, root_dir, (CHAR16*) L"\\loader\\keys\\auto");
+
+                if (!EFI_ERROR(err))
+                        return EFI_SUCCESS;
```

Comment:
if \loader\keys\auto doesn't exist we should continue silently here. i.e. check for the EFI version of ENOENT here, and skip the log message below